drivers/sx126x/sx126x_radio_hal.c

```diff
@@ -0,0 +1,724 @@
+                }
+                /* In case the packet is an ACK and the ACK filter is disabled,
+                 * indicate the frame reception */
+                else if (is_ack && !cfg.ack_filter) {
+                    DEBUG("[sx126x] Received ACK.\n");
+                    hal->cb(hal, IEEE802154_RADIO_INDICATION_RX_DONE);
+                }
+                /* If all failed, simply drop the frame and continue listening
+                 * to incoming frames */
+                else {
+                    DEBUG("[sx126x] Addr filter failed or ACK filter on.\n");
+                    _set_state(dev, STATE_RX);
+                   
+                }
+    }
+    else if (irq_mask & SX126X_IRQ_PREAMBLE_DETECTED) {
+        DEBUG("[sx126x] netdev: SX126X_IRQ_PREAMBLE_DETECTED\n");
+    }
+    else if (irq_mask & SX126X_IRQ_SYNC_WORD_VALID) {
+        DEBUG("[sx126x] netdev: SX126X_IRQ_SYNC_WORD_VALID\n");
+        
+    }
+    else if (irq_mask & SX126X_IRQ_HEADER_VALID) {
+        DEBUG("[sx126x] netdev: SX126X_IRQ_HEADER_VALID\n");
+        hal->cb(hal, IEEE802154_RADIO_INDICATION_RX_START);
+    }
+    else if (irq_mask & SX126X_IRQ_HEADER_ERROR) {
+        DEBUG("[sx126x] netdev: SX126X_IRQ_HEADER_ERROR\n");
+    }
+    else if (irq_mask & SX126X_IRQ_CRC_ERROR) {
+        DEBUG("[sx126x] netdev: SX126X_IRQ_CRC_ERROR\n");
+        hal->cb(hal, IEEE802154_RADIO_INDICATION_CRC_ERROR);
+    }
+    else if (irq_mask & SX126X_IRQ_CAD_DONE) {
+        if (irq_mask & SX126X_IRQ_CAD_DETECTED){
+            DEBUG("[sx126x] netdev: SX126X_IRQ_CAD_DETECTED \n");
+            dev->cad_detected = true;
+        }
+        DEBUG("[sx126x] netdev: SX126X_IRQ_CAD_DONE\n");
+        hal->cb(hal, IEEE802154_RADIO_CONFIRM_CCA);
+        _set_state(dev, STATE_RX);
+        
+    }
+    else if (irq_mask & SX126X_IRQ_TIMEOUT) {
+        DEBUG("[sx126x] netdev: SX126X_IRQ_TIMEOUT\n");
+    }
+    else {
+        DEBUG("[sx126x] netdev: SX126X_IRQ_NONE\n");
+    }
+      #if IS_USED(MODULE_SX126X_STM32WL)
+        NVIC_EnableIRQ(SUBGHZ_Radio_IRQn);
+    #endif  
+
+}
+}
+
+                        /*RADIO HAL FUNCTIONS*/
+
+static int _write(ieee802154_dev_t *hal, const iolist_t *iolist){
+
+    sx126x_t *dev = hal->priv;
+    (void)dev;
+    size_t pos = 0;
+    uint16_t chksum = 0;
+    /* Full buffer used for Tx */
+    sx126x_set_buffer_base_address(dev, 0x00, 0x00);
+    /* Write payload buffer */
+    for (const iolist_t *iol = iolist; iol; iol = iol->iol_next) {
+        if (iol->iol_len > 0) {
+            sx126x_write_buffer(dev, pos, iol->iol_base, iol->iol_len);
+            memcpy((&txbuf[pos]), iol->iol_base, iol->iol_len);
+            DEBUG("[sx126x] netdev: send: wrote data to payload buffer.\n");
+            chksum = ucrc16_calc_le(iol->iol_base, iol->iol_len, UCRC16_CCITT_POLY_LE, chksum);
+            pos += iol->iol_len;
+        }
+    }
+    
+    /* Ignore send if packet size is 0 */
+    if (!pos) {
+        return 0;
```

Comment:
the Radio HAL assumes the packet is valid (e.g not null).  Therefore you can skip these lines